### PR TITLE
Adding migrations with Alembic

### PR DESCRIPTION
**Pull request by HarsheetKakar**
Fixes Implement migrations with Alembic #349

### Checklist
- [x] My branch is up-to-date with upstream/develop branch.
- [ ] Everything works and tested for Python 3.5.2 and above.

### Description
Alembic database migration tool added for future migrations in database
### Change logs

- Alembic in requirements.txt
- set up for Alembic for single database

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -110,5 +110,5 @@ ENV/
 # Keys
 *.pem
 
-# Pipenv generated
-Pipfile*
\ No newline at end of file
+# Pipenv generated
```
Comment (de-sh) [inline .gitignore]: Undo changes to this file

**Comment by de-sh** [comment]
@HarsheetKakar the PEP8 tests say that the following issues exist in the file `hydrus/data/db_models.py`
```
14:89: E261 at least two spaces before inline comment
14:90: E262 inline comment should start with '# '
14:101: E501 line too long (131 > 100 characters)
```

File: hydrus/data/db_models.py [hunk 2]
```diff
@@ -8,9 +8,12 @@
 from typing import Any
 import datetime
 import uuid
+import os
 # from hydrus.settings import DB_URL
 
-engine = create_engine('sqlite:///database.db')
+BASE_DIR = os.path.dirname(os.path.dirname(os.path.dirname(os.path.realpath(__file__)))) #to put database file in the rootdirectory
```
Comment (de-sh) [inline hydrus/data/db_models.py]: This is where all issues occur. Our PEP8 test fails at 101 chars per line. Place the comment on top of the code.

Comment (HarsheetKakar) [reply hydrus/data/db_models.py]: done, thanks for pointing out

**Comment by Mec-iS** [comment]
we need documentation on how to run migrations

**Comment by HarsheetKakar** [comment]
> we need documentation on how to run migrations

please check if migrations.md works for our purposes

File: migrations.md [hunk 3]
```diff
@@ -3,24 +3,30 @@
...
-config.set_main_option('sqlalchemy.url', 'sqlite:///{}/database.db'.format(DATABASE_DIR))
-'''
+
+    DATABASE_DIR = <location of database.db>
+    config.set_main_option('sqlalchemy.url','sqlite:///{}/database.db'.format(DATABASE_DIR))
+
+
+
 
 Now you can make changes in db_model.
```
Comment (Guttz) [inline migrations.md]: Currently hydrus creates db_model with the assistance of doc_parse based on the Hydra API provided by the user.  That said, I don't think that's quite the use case we would have in the Hydra Ecosystem. I'll comment below how I personally think that could be done.

**Comment by Guttz** [comment]
So as a useful feature to be added for hydrus I can see the migration provided in two ways:

1. Instruct the user to manually run hydrus once(which means creating a copy of the db, update Hydra Doc and run hydrus) and so with the updated db_models then run alembic migrate with the old copy of the db.. Which would be basically using alembic by itself

2. Create a hydrus option (hydrus --migrate) that calls a script that does the above: update db_models based on Hydra Doc, run it with the database and outputs the upgrade script - a bonus nice addition would to potentially try to run the migration automatically. In my mind a nice flow for the user would be the following: 

image

**Comment by HarsheetKakar** [comment]
> So as a useful feature to be added for hydrus I can see the migration provided in two ways:
> 
> 1. Instruct the user to manually run hydrus once(which means creating a copy of the db, update Hydra Doc and run hydrus) and so with the updated db_models then run alembic migrate with the old copy of the db.. Which would be basically using alembic by itself
> 2. Create a hydrus option (hydrus --migrate) that calls a script that does the above: update db_models based on Hydra Doc, run it with the database and outputs the upgrade script - a bonus nice addition would to potentially try to run the migration automatically. In my mind a nice flow for the user would be the following:
> 
> image

1. can be done, I can just add it above env.py part
2. for this we need someone who knows click, im currently not working with it

**Comment by Mec-iS** [comment]
@Guttz please pick this up and provide a complete patch, You can just fork and open a new PR.

**Comment by Guttz** [comment]
New PR about migrations at https://github.com/HTTP-APIs/hydrus/pull/389.

**Pull request closed without merging**